src/dragonball/src/dbs_pci/src/vfio.rs

```diff
@@ -0,0 +1,1918 @@
+const PCI_CONFIG_IO_BAR: u32 = 0x1;
+// Memory BAR flags (lower 4 bits).
+const PCI_CONFIG_MEMORY_BAR_FLAG_MASK: u32 = 0xf;
+// 64-bit memory bar flag.
+const PCI_CONFIG_MEMORY_BAR_64BIT: u32 = 0x4;
+// Number of BARs for a PCI device
+const BAR_NUMS: u32 = 6;
+// PCI Header Type register index
+const PCI_HEADER_TYPE_REG_INDEX: u32 = 3;
+// First BAR register index
+const PCI_CONFIG_BAR0_INDEX: u32 = 4;
+// PCI ROM expansion BAR register index
+const PCI_ROM_EXP_BAR_INDEX: u32 = 12;
+// PCI interrupt pin and line register index
+const PCI_INTX_REG_INDEX: u32 = 15;
+const PAGE_SIZE: u64 = 4096;
+pub const VENDOR_NVIDIA: u16 = 0x10de;
+
+#[derive(Debug, thiserror::Error)]
+pub enum VfioPciError {
+    #[error("failed to create BAR {0}: {1}")]
+    CreateBar(u32, crate::Error),
+    #[error("device manager error: {0:?}")]
+    DeviceManager(#[source] dbs_device::device_manager::Error),
+    #[error("invalid assigned resources for VFIO PCI device")]
+    InvalidResources,
+    #[error("pci_vfio internal error")]
+    InternalError,
+    #[error("failed to set interrupt mode: {0:?}")]
+    InterruptManager(#[source] std::io::Error),
+    #[error("failed to map VFIO PCI region into guest: {0:?}")]
+    MapRegionGuest(#[source] kvm_ioctls::Error),
+    #[error("failed to mmap PCI device region: {0:?}")]
+    Mmap(#[source] std::io::Error),
+    #[error("failed to issue VFIO ioctl: {0:?}")]
+    Vfio(#[source] vfio_ioctls::VfioError),
+    #[error("failed to manager vm-pci device: {0:?}")]
+    VmPciError(#[source] crate::Error),
+}
+
+type Result<T> = std::result::Result<T, VfioPciError>;
+
+#[derive(Copy, Clone)]
+enum PciVfioSubclass {
+    VfioSubclass = 0xff,
+}
+
+impl PciSubclass for PciVfioSubclass {
+    fn get_register_value(&self) -> u8 {
+        *self as u8
+    }
+}
+
+#[derive(PartialEq)]
+pub(crate) struct VfioMsi {
+    state: MsiState,
+    cap: MsiCap,
+    cap_offset: u32,
+}
+
+impl VfioMsi {
+    fn with_in_range(&self, offset: u32) -> Option<u32> {
+        if offset >= self.cap_offset && offset < self.cap_offset + self.cap.size() {
+            Some(offset - self.cap_offset)
+        } else {
+            None
+        }
+    }
+}
+
+#[derive(PartialEq)]
+pub(crate) struct VfioMsix {
+    state: MsixState,
+    cap: MsixCap,
+    cap_offset: u32,
+    table_bir: u32,
+    table_offset: u64,
+    table_size: u64,
+    pba_bir: u32,
+    pba_offset: u64,
+    pba_size: u64,
+}
+
+impl VfioMsix {
+    fn with_in_range(&self, offset: u32) -> Option<u32> {
+        if offset >= self.cap_offset && offset < self.cap_offset + self.cap.len() as u32 {
+            Some(offset - self.cap_offset)
+        } else {
+            None
+        }
+    }
+}
+
+struct Interrupt {
+    vfio_dev: Arc<VfioDevice>,
+    irq_manager: Option<DeviceInterruptManager<Arc<KvmIrqManager>>>,
+    resources: DeviceResources,
+    has_legacy_irq: bool,
+    legacy_enabled: bool,
+    msi_enabled: bool,
+    msix_enabled: bool,
```

Comment:
this is necessary because `msi_enable` mean we have the *ability* to use msi but it is possible that msi resource is not allocated so it is None by now. and `msi` mean the *actual resource* of msi.

Reply:
also, comments are added.